### PR TITLE
feat: add maintenance mode

### PR DESCRIPTION
**Pull request by zyy17**
## What's changed

1. **Implement `Apply()` in `DatanodeDeployer` to handle maintenance mode**

   - When the original Pod needs to be restarted(for example, updated by admin) and the cluster uses remote WAL, the deployer will request metasrv to turn on maintenance mode before resource sync;
   - When the sync process is finished, the deployer will request metasrv again to turn on the maintenance mode;

2. **Add the validation of `EnableRegionFailover`**

   - When the cluster doesn't set up the remote WAL, the default `EnableRegionFailover` should be `false`;
   - When the cluster set up the remote WAL, the default `EnableRegionFailover` should be `false`;

## Summary by CodeRabbit

- **New Features**
	- Introduced maintenance mode for data nodes, including methods to manage and request maintenance from services.
	- Added the ability to handle region failover based on the presence of a remote WAL provider.

- **Enhancements**
	- Improved validation for region failover settings to ensure proper configuration.

- **Bug Fixes**
	- Fixed handling of the `EnableRegionFailover` field to allow null values, ensuring more flexible configuration management.

- **Tests**
	- Added tests to verify the default setting of `EnableRegionFailover`.

### CODE REVIEW

File: controllers/greptimedbcluster/controller.go [hunk 1]
```diff
@@ -301,6 +301,11 @@ func (r *Reconciler) validate(ctx context.Context, cluster *v1alpha1.GreptimeDBC
 		if err := r.validateTomlConfig(cluster.Spec.Meta.Config); err != nil {
 			return fmt.Errorf("invalid meta toml config: %v", err)
 		}
+		if cluster.Spec.Meta.EnableRegionFailover != nil && *cluster.Spec.Meta.EnableRegionFailover {
+			if cluster.Spec.RemoteWalProvider == nil {
+				return fmt.Errorf("remote wal provider must be specified when enable region failover")
+			}
+		}
 	}
```
Comment (daviderli614) [inline controllers/greptimedbcluster/controller.go:309]: suggestion: Only when RemoteWalProvider is turned on need to check EnableRegionFailover ？
```
if cluster.Spec.RemoteWalProvider  != nil {
  if !*cluster.Spec.Meta.EnableRegionFailover {
    return fmt.Errorf("remote wal provider must be specified when enable region failover")
  }
}
```

File: controllers/greptimedbcluster/deployers/datanode.go [hunk 2]
```diff
@@ -153,6 +257,57 @@ func (d *DatanodeDeployer) deleteStorage(ctx context.Context, cluster *v1alpha1.
...
 }
 
+// isPodRestart checks if the pod needs to be restarted. For convenience, we only compare the necessary fields.
+// TODO(zyy17): Do we have a easy way to implement this?
+func (d *DatanodeDeployer) isPodRestart(new, old appsv1.StatefulSet) bool {
+	var (
+		newPodTemplate = new.Spec.Template
+		oldPodTemplate = old.Spec.Template
+	)
+
```
Comment (daviderli614) [inline controllers/greptimedbcluster/deployers/datanode.go:267]: question：By checking whether old.spec is equal to new.spec, we can determine whether to restart?

File: controllers/greptimedbcluster/deployers/datanode.go [hunk 3]
```diff
@@ -153,6 +257,57 @@ func (d *DatanodeDeployer) deleteStorage(ctx context.Context, cluster *v1alpha1.
...
+	// If the tolerations, affinity, nodeSelector are changed, the original Pod may need to be restarted for re-scheduling.
+	if !reflect.DeepEqual(newPodTemplate.Spec.Tolerations, oldPodTemplate.Spec.Tolerations) ||
+		!reflect.DeepEqual(newPodTemplate.Spec.Affinity, oldPodTemplate.Spec.Affinity) ||
+		!reflect.DeepEqual(newPodTemplate.Spec.NodeSelector, oldPodTemplate.Spec.NodeSelector) {
+		return true
+	}
+
+	// Compare the main container settings.
+	newMainContainer := newPodTemplate.Spec.Containers[constant.MainContainerIndex]
+	oldMainContainer := oldPodTemplate.Spec.Containers[constant.MainContainerIndex]
```
Comment (daviderli614) [inline controllers/greptimedbcluster/deployers/datanode.go:285]: question：By checking whether old.spec is equal to new.spec, we can determine whether to restart?

File: controllers/greptimedbcluster/controller.go (hunk 1 above)
Comment (zyy17) [reply controllers/greptimedbcluster/controller.go:309]: make sense.

Comment (zyy17) [reply controllers/greptimedbcluster/controller.go:309]: When `EnableRegionFailover` is true, the user have to use remote wal. Howevevr, when you use remote wal, you don't have to enable region failover.

File: controllers/greptimedbcluster/deployers/datanode.go (hunk 3 above)
Comment (zyy17) [reply controllers/greptimedbcluster/deployers/datanode.go:285]: We don't have to turn on maintenance mode when scaling. Maybe `isPodRestart()` is not a appropriate name, let me change to `isOldPodRestart()`.

**Review by daviderli614 — APPROVED** [review]

**Pull request merged**